File: benchmark-framework-source/src/main/java/com/benchmark/core/BenchmarkConfig.java
package com.benchmark.core;

import java.util.Arrays;
import java.util.List;

/**
 * Immutable configuration for a benchmark run.
 * Build via {@link BenchmarkConfig.Builder}.
 */
public final class BenchmarkConfig {

    private final List<Integer> concurrencyLevels;
    private final int metricsSampleIntervalMs;
    private final String outputDirectory;
    private final boolean generateHtmlReport;
    private final boolean generateCharts;
    private final boolean verbose;
    private final int taskTimeoutSeconds;

    private BenchmarkConfig(Builder b) {
        this.concurrencyLevels        = b.concurrencyLevels;
        this.metricsSampleIntervalMs  = b.metricsSampleIntervalMs;
        this.outputDirectory          = b.outputDirectory;
        this.generateHtmlReport       = b.generateHtmlReport;
        this.generateCharts           = b.generateCharts;
        this.verbose                  = b.verbose;
        this.taskTimeoutSeconds       = b.taskTimeoutSeconds;
    }

    public List<Integer> getConcurrencyLevels()  { return concurrencyLevels; }
    public int getMetricsSampleIntervalMs()       { return metricsSampleIntervalMs; }
    public String getOutputDirectory()            { return outputDirectory; }
    public boolean isGenerateHtmlReport()         { return generateHtmlReport; }
    public boolean isGenerateCharts()             { return generateCharts; }
    public boolean isVerbose()                    { return verbose; }
    public int getTaskTimeoutSeconds()            { return taskTimeoutSeconds; }

    // -------------------------------------------------------------------------

    public static Builder builder() { return new Builder(); }

    public static final class Builder {

        private List<Integer> concurrencyLevels       = Arrays.asList(1, 2, 4, 8, 16, 32);
        private int           metricsSampleIntervalMs = 200;
        private String        outputDirectory         = "benchmark-results";
        private boolean       generateHtmlReport      = true;
        private boolean       generateCharts          = true;
        private boolean       verbose                 = false;
        private int           taskTimeoutSeconds      = 300;

        /**
         * Thread counts at which each task will be exercised.
         * E.g. {@code concurrencyLevels(1, 2, 4, 8, 16, 32)}
         */
        public Builder concurrencyLevels(Integer... levels) {
            this.concurrencyLevels = Arrays.asList(levels);
            return this;
        }

        /**
         * How often (ms) the metrics collector samples CPU and memory.
         * Lower = more accurate averages, higher CPU overhead of collection.
         */
        public Builder metricsSampleIntervalMs(int ms) {
            this.metricsSampleIntervalMs = ms;
            return this;
        }

        /** Directory where PNGs, JSON, and the HTML report are written. */
        public Builder outputDirectory(String dir) {
            this.outputDirectory = dir;
            return this;
        }

        public Builder generateHtmlReport(boolean b) {
            this.generateHtmlReport = b;
            return this;
        }

        public Builder generateCharts(boolean b) {
            this.generateCharts = b;
            return this;
        }

        /** Print per-iteration timings to stdout. */
        public Builder verbose(boolean b) {
            this.verbose = b;
            return this;
        }

        /** Hard wall-clock cap per concurrency level to prevent hangs. */
        public Builder taskTimeoutSeconds(int s) {
            this.taskTimeoutSeconds = s;
            return this;
        }

        public BenchmarkConfig build() {
            if (concurrencyLevels == null || concurrencyLevels.isEmpty())
                throw new IllegalStateException("At least one concurrency level is required");
            if (metricsSampleIntervalMs < 10)
                throw new IllegalStateException("Sample interval must be >= 10 ms");
            return new BenchmarkConfig(this);
        }
    }
}

File: python/benchmark/core/benchmark_config.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Sequence


@dataclass(frozen=True)
class BenchmarkConfig:
    """Immutable configuration for a benchmark run."""

    concurrency_levels: tuple[int, ...] = (1, 2, 4, 8, 16, 32)
    metrics_sample_interval_ms: int = 200
    output_directory: str = "benchmark-results"
    generate_html_report: bool = True
    generate_charts: bool = True
    verbose: bool = False
    task_timeout_seconds: int = 300

    @staticmethod
    def builder() -> "BenchmarkConfig.Builder":
        return BenchmarkConfig.Builder()

    class Builder:
        def __init__(self) -> None:
            self._concurrency_levels: tuple[int, ...] = (1, 2, 4, 8, 16, 32)
            self._metrics_sample_interval_ms: int = 200
            self._output_directory: str = "benchmark-results"
            self._generate_html_report: bool = True
            self._generate_charts: bool = True
            self._verbose: bool = False
            self._task_timeout_seconds: int = 300

        def concurrency_levels(self, *levels: int) -> "BenchmarkConfig.Builder":
            self._concurrency_levels = tuple(levels)
            return self

        def metrics_sample_interval_ms(self, ms: int) -> "BenchmarkConfig.Builder":
            self._metrics_sample_interval_ms = ms
            return self

        def output_directory(self, dir_: str) -> "BenchmarkConfig.Builder":
            self._output_directory = dir_
            return self

        def generate_html_report(self, b: bool) -> "BenchmarkConfig.Builder":
            self._generate_html_report = b
            return self

        def generate_charts(self, b: bool) -> "BenchmarkConfig.Builder":
            self._generate_charts = b
            return self

        def verbose(self, b: bool) -> "BenchmarkConfig.Builder":
            self._verbose = b
            return self

        def task_timeout_seconds(self, s: int) -> "BenchmarkConfig.Builder":
            self._task_timeout_seconds = s
            return self

        def build(self) -> "BenchmarkConfig":
            if not self._concurrency_levels:
                raise ValueError("At least one concurrency level is required")
            if self._metrics_sample_interval_ms < 10:
                raise ValueError("Sample interval must be >= 10 ms")
            return BenchmarkConfig(
                concurrency_levels=self._concurrency_levels,
                metrics_sample_interval_ms=self._metrics_sample_interval_ms,
                output_directory=self._output_directory,
                generate_html_report=self._generate_html_report,
                generate_charts=self._generate_charts,
                verbose=self._verbose,
                task_timeout_seconds=self._task_timeout_seconds,
            )
